app/posts/finishing-touches-for-forms.md

```diff
@@ -0,0 +1,103 @@
+
+This causes a few problems.
+
+The GOV.UK design system includes some complex rules around linking to fields from the error summary. For example, errors on checkbox or radio button fields should link to the first checkbox or radio button.
+
+Rails doesn't support this and for it to work the hyperlinks and targets need to be consistent. Because forms can contain custom logic, it's vital that the form builder is responsible for generating both.
+
+We could solve the problem by wrapping the whole page in a `<form>` element, which would allow us to place the error summary wherever we like, but it's a hack. It isn't semantically correct and is likely to lead to other problems.
+
+Thankfully, Rails has our backs. We can use [`content_for`](https://guides.rubyonrails.org/layouts_and_rendering.html#using-the-content-for-method) to insert content into a named block in our layout. If we add a named block called `top_of_main` at the top of our `<main>` element, like this:
+
+```html
+<div class="govuk-width-container">
+  <main class="govuk-main-wrapper" id="main-content">
+    <%= yield :top_of_main %>
+    <h1 class="govuk-heading-xl">Default page template</h1>
+
+    <!-- the rest of our page -->
+```
+
+We can send our error summary from the form to it, like this:
+
+```html
+<%= form_with(model: @object, url: some_path) do |form| %>
+  <%= content_for(:top_of_main) { form.govuk_error_summary } %>
+
+  <%= form.govuk_text_field(:how_many_juggling_items) %>
+<% end %>
+```
+
+And now our error summary is in the right spot.
+
+![Two screenshots side by side showing a form with errors. The left one has the error summary after the h1 and the right one has the error summary first](/assets/posts/finishing-touches-for-forms/error-summary-side-by-side.png)
+
+## Prefixing the page title with `Error:`
+
+When a form submission results in a validation failure we need to make it clear to the user that something is wrong.
+
+[The guidance says](https://design-system.service.gov.uk/patterns/validation/#how-to-tell-the-user-about-validation-errors):
+
+> Add ‘Error: ’ to the beginning of the page `<title>` so screen readers read it out as soon as possible
+
+This is a little tricky because we need a some logic that can determine whether there's an error on the page, and we set the `<title>` in the document `<head>` before we've rendered the form in the `<body>`.
+
+We can use `content_for` to our advantage here, too.
+
+First, we set page titles using a [helper method](https://www.rubyguides.com/2020/01/rails-helpers/) and pass in the text and value that represents whether or not there are any errors on the page:
+
+```html
+<%
+  page_title(
+    "How many implements can you juggle with?",
+    error: @object.errors.any?
+  )
+%>
+```
+
+If there are errors we can add the 'Error:' prefix:
+
+```ruby
+def page_title(text, error: false)
```

Comment:
Any reason not to recommend our existing helper here? https://govuk-components.netlify.app/helpers/title-with-error-prefix/ 😄

Reply:
🤦‍♂️

Oh dear, I forgot it was there. I'll amend the post 😅 Thank you!

Reply:
Haha, doh! I’m not sure it’s yet widely used in DfE repos as it was added after most of them already had their own helper?